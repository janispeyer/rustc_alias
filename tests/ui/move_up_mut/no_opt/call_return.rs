// compile-flags: -Zmir-emit-retag

fn read() -> &'static mut i32 {
    Box::leak(Box::new(42))
}

fn call_return(mut x: &mut i32) -> i32 {
    *x = 7;
    x = read();
    return *x;
}

fn main() {
    let mut x = 1;
    call_return(&mut x);
}
