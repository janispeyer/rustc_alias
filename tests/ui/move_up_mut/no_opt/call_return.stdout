# Analysing main
# Analysing call_return
# MaybeTopOfBorrowStack Analysis
bb0
[ 0] [_1] <- Retag([fn entry] _1)
[ 1] [_1] <- (*_1) = const 7_i32
[ 2] [_1] <- StorageLive(_3)
[ 3] [_1] <- StorageLive(_4)
[ 4] [_1] <- _4 = read() -> bb1

bb1
[ 0] [_1] <- Retag(_4)
[ 1] [_1] <- _3 = &mut (*_4)
[ 2] [_1] <- Retag(_3)
[ 3] [] <- _1 = move _3
[ 4] [] <- Retag(_1)
[ 5] [] <- StorageDead(_3)
[ 6] [] <- StorageDead(_4)
[ 7] [] <- _0 = (*_1)
[ 8] [] <- return

# FindImmutabilitySpans Analysis
bb0
[ 0] {} <- Retag([fn entry] _1)
[ 1] {_1: Span(bb0[1])} <- (*_1) = const 7_i32
[ 2] {_1: Span(bb0[1])} <- StorageLive(_3)
[ 3] {_1: Span(bb0[1])} <- StorageLive(_4)
[ 4] {_1: Span(bb0[1])} <- _4 = read() -> bb1

bb1
[ 0] {_1: Span(bb0[1])} <- Retag(_4)
[ 1] {_1: Span(bb0[1])} <- _3 = &mut (*_4)
[ 2] {_1: Span(bb0[1])} <- Retag(_3)
[ 3] {_1: Top} <- _1 = move _3
[ 4] {_1: Top} <- Retag(_1)
[ 5] {_1: Top} <- StorageDead(_3)
[ 6] {_1: Top} <- StorageDead(_4)
[ 7] {_1: Top} <- _0 = (*_1)
[ 8] {_1: Top} <- return


# CFG Before
bb0
Retag([fn entry] _1)
(*_1) = const 7_i32
StorageLive(_3)
StorageLive(_4)
_4 = read() -> bb1

bb1
Retag(_4)
_3 = &mut (*_4)
Retag(_3)
_1 = move _3
Retag(_1)
StorageDead(_3)
StorageDead(_4)
_0 = (*_1)
return

# CFG After
bb0
Retag([fn entry] _1)
(*_1) = const 7_i32
StorageLive(_3)
StorageLive(_4)
_4 = read() -> bb1

bb1
Retag(_4)
_3 = &mut (*_4)
Retag(_3)
_1 = move _3
Retag(_1)
StorageDead(_3)
StorageDead(_4)
_0 = (*_1)
return

# Analysing read
